### PR TITLE
Jetpack Manage: implement license bundle selection

### PR DESCRIPTION
**Pull request by yashwin**
Resolves https://github.com/Automattic/jetpack-genesis/issues/92
Resolves https://github.com/Automattic/jetpack-genesis/issues/93

## Proposed Changes

This PR adds

- IssueLicenseContext context API to manage the issue license states
- Implements the license bundle selection

## Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Visit the Jetpack Cloud link > Click `Licenses` > Click the `Issue New License` button and append the URL with `?flags=jetpack/bundle-licensing` and verify that you can see the newly created component being rendered
2. Select the single license & bundle licenses and verify that the count is incremented based on the selection. Also, verify that you can select a license from different bundle sizes simultaneously and that the selected one is shown when returning to the selected bundle tab. Please ignore the duplicate CTA buttons. They will need to be refactored and fixed and will be done in another PR.
3. Also, verify that the issue license works as expected without `?flags=jetpack/bundle-licensing`.

Screenshot 2023-11-20 at 10 08 11 AM

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/jetpack-cloud/sections/partner-portal/issue-license-v2/index.tsx [hunk 1]
```diff
@@ -58,7 +65,13 @@ export default function IssueLicenseV2() {
 			</LayoutTop>
 
 			<LayoutBody>
-				<div>body here</div>
+				<IssueLicenseContext.Provider value={ { setSelectedLicenses, selectedLicenses } }>
+					<IssueMultipleLicensesForm
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/issue-license-v2/index.tsx]: Hey @yashwin, I just wanted to confirm if this is the component for building our new product listing. I have some concerns that we might end up increasing the complexity of the current component instead of creating a separate one for the new product listing. As this project is still in the testing phase, it would be easier for us to revert the changes if needed. wdyt?

File: client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx [hunk 2]
```diff
@@ -115,8 +125,37 @@ export default function IssueMultipleLicensesForm( {
...
+	const handleSelectBundleLicense = useCallback(
+		( product: APIProductFamilyProduct ) => {
+			const productBundle = {
+				...product,
+				quantity,
+			};
+			const index = selectedLicenses.findIndex(
+				( item ) => item.quantity === productBundle.quantity && item.slug === productBundle.slug
+			);
+			if ( index !== -1 ) {
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx]: [Non-blocker] I would encourage to make this a positive condition and make it more easier to reason about.

```
            if ( index === -1 ) {
				// Item doesn't exist, add it
				setSelectedLicenses( [ ...selectedLicenses, productBundle ] );
			} else {
				// Item exists, remove it
				setSelectedLicenses( selectedLicenses.filter( ( _, i ) => i !== index ) );
			}
```

**Review by jkguidaven — APPROVED** [review]
Nice work on adding support with mixed purchases. I can confirm it works as expected. I have a few comments, but those are non-blocking suggestions.

I would like to encourage the design team to review the mix purchase capabilities and ensure that there are no changes in the UI design.

**Comment by yashwin** [comment]
Thanks for the review, @jkguidaven!

I agree with your suggestions. I have made the changes: https://github.com/Automattic/wp-calypso/pull/84330/commits/59dd2ef436c41497d0b291e99b56be3672c97564

> I would like to encourage the design team to review the mix purchase capabilities and ensure that there are no changes in the UI design.

I have already asked for the design review 👍

File: client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx [hunk 3]
```diff
@@ -28,8 +28,8 @@ import {
 	getSelectedProductSlugs,
 } from 'calypso/state/partner-portal/products/selectors';
 import { APIProductFamilyProduct, PartnerPortalStore } from 'calypso/state/partner-portal/types';
+import useSubmitForm from '../issue-license-v2/hooks/use-submit-form';
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx:31]: I suggest keeping this file in the same directory to minimize the number of files affected by this PR.

File: client/jetpack-cloud/sections/partner-portal/issue-license-v2/licenses-form/index.tsx [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+		.map( ( license ) => license.quantity )
+		.reduce( ( a, b ) => a + b, 0 );
+
+	const isSelected = ( slug: string ) =>
+		selectedLicenses.findIndex(
+			( license ) => license.slug === slug && license.quantity === quantity
+		) !== -1;
+
+	return (
+		<div className="issue-multiple-licenses-form">
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/issue-license-v2/licenses-form/index.tsx]: We should probably change the class name to match the new component name. wdyt?

**Review by jkguidaven — COMMENTED** [review]
Thanks for taking action on my previous comments. 🙏 

I have some additional comments on the new changes. Let me know what you think.

File: client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx (hunk 3 above)
Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx:31]: We will eventually get rid of this directory once we ship this project during the cleanup. That's why I moved it here and copied the styles over.

File: client/jetpack-cloud/sections/partner-portal/issue-license-v2/licenses-form/index.tsx (hunk 4 above)
Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/issue-license-v2/licenses-form/index.tsx]: Agree

**Comment by yashwin** [comment]
@jkguidaven, thanks for your suggestions. I have made the changes accordingly: https://github.com/Automattic/wp-calypso/pull/84330/commits/cb11ebb31bcc3ad056af094f35758d400a405724. Could you please take another look? Thanks! 🙂

**Review by jkguidaven — APPROVED** [review]
Thanks for addressing all my PR comments. LGTM 👍 nice work

**Comment by jeffgolenski** [comment]
@yashwin Good start my friend. I'm unsure what aspects are being broken into different PRs, but they'll need to all come together to see everything working together. 

Just a few notes. I recorded a video for ya! https://www.loom.com/share/56ee344feb244c82b14080422d3532a1

* Mixed bundle purchases looks good. I kept that in mind when designing. We just need to ensure the 
* Total label doesn't seem to be showing in this PR, is that intentional?
* There are two issue license buttons showing. We just need the dynamic one to show up when licenses bundles are selected. 

Let me know if there are any questions, or other PRs to review! Thanks my friend.

**Comment by yashwin** [comment]
> @yashwin Good start my friend. I'm unsure what aspects are being broken into different PRs, but they'll need to all come together to see everything working together.

Thanks for the review & feedback, @jeffgolenski! Yes, this PR mainly addresses the mixed bundle purchases. 

> * Mixed bundle purchases looks good. I kept that in mind when designing.

👍 

> * Total label doesn't seem to be showing in this PR, is that intentional?

Yes, this will be added in a PR later.

> * There are two issue license buttons showing. We just need the dynamic one to show up when licenses bundles are selected.

This is expected for now. We will fix it in an upcoming PR. I didn't want to increase the scope of this PR.
